Ignore empty and whitespace VM options

sharon-wang opened the pull request:
Currently, Openj9 breaks legacy applications running JVM from C++ with empty VM options, providing the error message: JVMJ9VM007E Command-line option unrecognised.

Instead, these empty options should be ignored, to match the RI behaviour.

Fixes: #5332. 

Note: this PR addresses ignoring options that are empty or all whitespace, but continues to reject ASCII characters beyond 0x7F.

Signed-off-by: Sharon Wang <sharon-wang-cpsc@outlook.com>

pshipton commented:
jenkins test sanity plinux jdk11

pshipton commented:
jenkins test extended plinux jdk11

pshipton commented:
We can ignore the extended test failures, the commit message includes `Command-line option unrecognised` which the test is looking for as a failure condition.

jenkins compile win jdk11

sharon-wang commented on runtime/vm/jvminit.c:
```diff
@@ -2421,6 +2426,16 @@ static UDATA checkArgsConsumed(J9PortLibrary* portLibrary, J9VMInitArgs* j9vm_ar
+/* Returns TRUE if a string is empty or if it contains only whitespace characters */
+static BOOLEAN isEmpty(const char * str) {
+	while('\0' != *str) {
+		if (0 == isspace((unsigned char) *str)) {
```
@JasonFengJ9 great question - I wasn't entirely sure about this, but here is my thinking:
`isspace(int c)` expects that c is representable as an `unsigned char` or is equal to `EOF`, otherwise the behaviour is undefined. I read that C/C++ accepts `char` to be signed or unsigned, depending on the platform and compiler. To ensure that `isspace()` receives an `unsigned char`, I added the cast. Should we expect to deal with either signed or unsigned `char`? If we know that we will only deal with unsigned, I can remove the cast.

JasonFengJ9 replied:
@sharon-wang good thinking! this `optionString` could be from outside of `JVM` just like the legacy application #5332, so technically it might contain signed char values which are not `ASCII` codes. 
It appears this PR is going to reject such `optionString` with `JVMJ9VM007E Command-line option unrecognised`, what's the `RI` behaviors?

pshipton replied:
@JasonFengJ9 do you have it backwards? OpenJ9 already rejects empty `optionString` with `JVMJ9VM007E Command-line option unrecognised`. This change is to ignore the empty options, to match the RI behavior.

JasonFengJ9 replied:
@pshipton right, this PR intends to ignore the empty options to match `RI` behaviors. From code perspective particularly inspired by this `unsigned char` cast, this PR is still going to reject those `signed char` options, my question was: does `RI` reject them as well or still ignore?

sharon-wang replied:
@JasonFengJ9 Compared behaviour between RI and OpenJ9 - RI ignores and this current implementation does in fact reject.

To deal with `signed char`, a check can be added to this function so that we break and return false if we encounter one. With this change, it would probably be better to rename this function to reflect that we are checking if we should ignore, rather than just an empty check. Would this be a reasonable approach?

JasonFengJ9 replied:
Chatted with @sharon-wang about this, we might need some help from JCL team to give a high level description about what `char` does `RI` ignore or reject.

pshipton replied:
I don't think we should be ignoring `signed chars`, i.e. chars > 0x7fff, if that is what you mean.

JasonFengJ9 replied:
Apparently `RI` ignores some `char` values `>0x7F`, need some hints from `JCL` to match `RI` behaviors.

pshipton replied:
What are you testing? It wouldn't be hard to use glaunch or some custom launcher and run 255 times, passing in each single char value to check the behavior. Although according to the spec, the options are originally passed in the platform encoding, so that would only tell us the behavior for the current encoding. By the time we get here, I'd expect the options to be in UTF8, although I've only found code for Windows that does the conversion.

JasonFengJ9 replied:
Testing all possible `char` values was suggested to Sharon initially though I felt it is kind of hack. In addition I was wondering if there is any possibility that a `char` string pattern is to be checked instead of just single `char` value. That's why I suggested to seek JCL advice for accurate `RI` behaviours description. 

As per https://github.com/eclipse/openj9/files/3035727/sampleJNI.zip, the optionString was passed in within following code snippet:
```
       JavaVMOption* options = new JavaVMOption[2];   // JVM invocation options
       options[0].optionString = "";     ---> this is what JVM received.
```
A non-ascii code could be passed as well like following:
```
char testString[10];
testString[0] = 0xA0;
testString[1] = 0x0;
options[0].optionString = testString;
```
It appears `RI` ignore this `optionString` as well but this PR doesn't. So we are looking for a better understanding about what can be ignored or rejected by `RI`.

DanHeidinga replied:
I don't think we need the cast as `isSpace` is declared to take an `int`.  Both signed and unsigned chars can be promoted to int.

Further, `isSpace(int)` handles the most likely space characters and using it allows this PR to solve the problem originally reported.  Let's not rathole down every possible character combination right now.  It's fine to follow up with the JCL team and target matching the RI for those additional cases in a future release.

Let's get this fix into master and the 0.14 branch if we can.

DanHeidinga commented on runtime/vm/jvminit.c:
```diff
@@ -2393,6 +2394,10 @@ static UDATA checkArgsConsumed(J9PortLibrary* portLibrary, J9VMInitArgs* j9vm_ar
 			char* envVar = j9vm_args->j9Options[i].fromEnvVar;
 
+			/* If the option string is empty or all whitespace */
+			if (TRUE == isEmpty(optString)) {
```
`isEmpty()` is already a boolean expression - no need to check `TRUE ==`

sharon-wang commented:
@DanHeidinga @JasonFengJ9 @pshipton 

Removed `(unsigned char)` cast, renamed function to `shouldIgnoreOption()` to be more inclusive (but perhaps this is a bit too general?)

Added note to PR description to mention that extended ASCII characters continue to be rejected and we are only addressing the empty/whitespace options at this time.

This should be good to go pending a final review.

sharon-wang commented:
@JasonFengJ9 Thanks for the feedback! Changed the function name back to `isEmpty()` since it better reflects what this PR adds.

DanHeidinga commented on runtime/vm/jvminit.c:
```diff
@@ -2392,7 +2393,11 @@ static UDATA checkArgsConsumed(J9PortLibrary* portLibrary, J9VMInitArgs* j9vm_ar
-
+			
+			/* If an option is empty or contains only whitespace, it should be ignored instead of rejected */
+			if (isEmpty(optString)) {
```
can `optString` be null here?  I see an explicit null check on it on line 2402 before calling `strncmp` on it...

sharon-wang replied:
Using https://github.com/eclipse/openj9/files/3035727/sampleJNI.zip and adding a `NULL` string to the args array `options[0].optionString = NULL;` results in a seg fault for both the RI and the J9 implementations.

However, it's probably best to add the explicit null check anyways - will be adding this in.

keithc-ca commented on runtime/vm/jvminit.c:
```diff
@@ -2392,7 +2393,11 @@ static UDATA checkArgsConsumed(J9PortLibrary* portLibrary, J9VMInitArgs* j9vm_ar
-
+			
+			/* If an option is empty or contains only whitespace, it should be ignored instead of rejected */
+			if (optString && isEmpty(optString)) {
```
Please be explicit with the NULL check: `(NULL != optString) && isEmpty(optString)`

keithc-ca commented:
I don't think options that are blank should be ignored: some applications need arguments that are entirely blank: e.g. `cut -d " "`

pshipton commented:
> I don't think options that are blank should be ignored

@keithc-ca this change is to ignore blank VM options (before the java class name), not the arguments passed to the java app

pshipton commented:
Added comments on the code, but not sure anybody sees them. The removal of the `(unsigned char)` in the call to isspace() seems wrong.

The man pages for the isspace() state `which must have the value of an unsigned char` or `the value of which the application shall ensure is a character representable as an unsigned char`.

sharon-wang commented:
@pshipton The `(unsigned char)` cast has been added back in. I'm not sure that I'm able to see the comments on the code.

keithc-ca commented on runtime/vm/jvminit.c:
```diff
@@ -2392,7 +2393,11 @@ static UDATA checkArgsConsumed(J9PortLibrary* portLibrary, J9VMInitArgs* j9vm_ar
-
+			
+			/* If an option is empty or contains only whitespace, it should be ignored instead of rejected */
+			if ((NULL != optString) && isEmpty(optString)) {
```
The reference implementation does not ignore empty options if `JavaVMInitArgs.ignoreUnrecognized` is false.

sharon-wang replied:
@keithc-ca Would it best to just add the `isEmpty()` check to the existing check for ignoreUnrecognized: https://github.com/eclipse/openj9/blob/9453b9844670a117cc67aca2aed5815bec8dc7bf/runtime/vm/jvminit.c#L2401-L2404
i.e.
`if (ignoreUnrecognized && optString && (isEmpty() || !strncmp(optString, "-X", 2) || *optString=='_'))`

sharon-wang replied:
I have pushed what I mentioned above, matching the reference implementation behaviour for when `JavaVMInitArgs.ignoreUnrecognized` is false.

pshipton commented:
We should add some testing for this. We have genlauncher.c in the test material already that can be leveraged. I'm ok to have the tests done in a separate PR to ensure we get this change delivered for the 0.14 release.

pshipton commented:
Is "" ignored only when ignoreUnrecognized is set, or always?

sharon-wang commented:
@pshipton `""` is ignored only when `ignoreUnrecognized` = true for RI

pshipton commented:
Before this gets merged, please open another issue for ignoring non-ascii whitespace in options.

Responding to an earlier code comment, note that 0xA0 is a "no-break space" in Unicode. Unicode can be used to determine what is whitespace. i.e. https://en.wikipedia.org/wiki/Unicode_character_property has a list.

optionString is specified as `the option as a string in the default platform encoding` i.e. if the default platform encoding is ISO8859-1 to -11, then 0xA0 is a 1-1 mapping to Unicode.

Since optionString is in the platform encoding (and we don't want to understand all the encodings), In order to skip non-ascii whitespace we'll have to convert the options to Unicode in UTF8 format, and then look at each UTF8 char to determine if it's whitespace. The conversion to Unicode isn't being done now (except it seems on Windows), and will introduce additional overhead. If we convert everything to Unicode/UTF8, we may want change other parts of the options processing and creation of system properties.

keithc-ca commented on runtime/vm/jvminit.c:
```diff
@@ -2392,9 +2393,10 @@ static UDATA checkArgsConsumed(J9PortLibrary* portLibrary, J9VMInitArgs* j9vm_ar
+			
+			/* If ignoreUnrecognized is set to JNI_TRUE, we should not reject any options that are: 
+				empty or contain only whitespace, or unrecognized options beginning with -X or _ */
+			if (ignoreUnrecognized && (NULL != optString) && (isEmpty(optString) || !strncmp(optString, "-X", 2) || *optString=='_')) {
```
This doesn't ignore `NULL`s which I thought was part of the goal (to avoid segfaults).

pshipton replied:
This was never the goal. Perhaps we should deal with that separately from this PR. Referring back to an earlier comment OpenJ9 and the RI have the same behavior `adding a NULL string to the args array options[0].optionString = NULL; results in a seg fault for both the RI and the J9 implementations.`

keithc-ca replied:
Perhaps that better left alone (or at least for another day). It might even be considered helpful to identify problems in the application attempting to launch a VM.

pshipton commented:
jenkins test sanity win jdk8

pshipton commented:
jenkins test extended plinux jdk8

sharon-wang commented:
@pshipton created issue: https://github.com/eclipse/openj9/issues/5387 for ignoring non-ASCII whitespace in options.

pshipton commented:
@sharon-wang please cherry pick the commit and create a new PR against the v0.14.0-release branch.

sharon-wang commented:
@pshipton PR created: https://github.com/eclipse/openj9/pull/5392
